sdk/webpubsub/azure-messaging-webpubsub-client/src/main/java/com/azure/messaging/webpubsub/client/WebPubSubAsyncClient.java

```diff
@@ -0,0 +1,176 @@
+    }
+
+    public Mono<WebPubSubResult> joinGroup(String group, long ackId) {
+        return Mono.fromCallable(() -> {
+            session.getBasicRemote().sendObject(new JoinGroupMessage().setGroup(group).setAckId(ackId));
+            return new WebPubSubResult();
+        }).subscribeOn(Schedulers.boundedElastic()).then(waitForAckMessage(ackId));
+    }
+
+    public Mono<WebPubSubResult> leaveGroup(String group) {
+        return leaveGroup(group, nextAckId());
+    }
+
+    public Mono<WebPubSubResult> leaveGroup(String group, long ackId) {
+        return Mono.fromCallable(() -> {
+            session.getBasicRemote().sendObject(new LeaveGroupMessage().setGroup(group).setAckId(ackId));
+            return new WebPubSubResult();
+        }).subscribeOn(Schedulers.boundedElastic()).then(waitForAckMessage(ackId));
+    }
+
+    public Mono<WebPubSubResult> sendMessageToGroup(String group, BinaryData content, WebPubSubDataType dataType) {
+        return sendMessageToGroup(group, content, dataType, nextAckId(), false, false);
+    }
+
+    public Mono<WebPubSubResult> sendMessageToGroup(String group, BinaryData content, WebPubSubDataType dataType,
+                                                    long ackId, boolean noEcho, boolean fireAndForget) {
+
+        Mono<WebPubSubResult> sendMono = Mono.fromCallable(() -> {
+            BinaryData data = content;
+            if (dataType == WebPubSubDataType.BINARY) {
+                data = BinaryData.fromBytes(Base64.getEncoder().encode(content.toBytes()));
+            }
+
+            session.getBasicRemote().sendObject(new SendToGroupMessage()
+                .setGroup(group)
+                .setData(data)
+                .setDataType(dataType.name().toLowerCase(Locale.ROOT))
+                .setAckId(ackId)
+                .setNoEcho(noEcho));
+            return (WebPubSubResult) null;
+        }).subscribeOn(Schedulers.boundedElastic());
+
+        if (!fireAndForget) {
+            sendMono = sendMono.then(waitForAckMessage(ackId));
+        } else {
+            sendMono = sendMono.then(Mono.just(new WebPubSubResult()));
+        }
+        return sendMono;
+    }
+
+    public Flux<GroupDataMessage> receiveGroupMessages() {
+        return messageSink.asFlux().filter(m -> m instanceof GroupDataMessage).cast(GroupDataMessage.class);
+    }
+
+    private static final AtomicLong ACK_ID = new AtomicLong(0);
+    private long nextAckId() {
+        return ACK_ID.getAndIncrement();
+    }
+
+    private Flux<AckMessage> receiveAckMessages() {
+        return messageSink.asFlux().filter(m -> m instanceof AckMessage).cast(AckMessage.class);
+    }
+
+    private Mono<WebPubSubResult> waitForAckMessage(long ackId) {
+        return receiveAckMessages()
+            .filter(m -> ackId == m.getAckId())
+            .map(m -> new WebPubSubResult(m.getAckId()))
+            .next()
+            // error handling
+            .switchIfEmpty(Mono.error(new RuntimeException()));
+    }
```

Comment:
TODO: we want to retry if network disrupted, we want to abort if client is stopped by user.